Implement derive macro for ToVariant and FromVariant

toasteater opened the pull request:
The derive macro does the following mapping between Rust structures and Godot types:

- `Newtype(inner)` is unwrapped to `inner`
- `Tuple(a, b, c)` is represented as a `VariantArray` (`[a, b, c]`)
- `Struct { a, b, c }` is represented as a `Dictionary` (`{ "a": a, "b": b, "c": c }`)
- `Unit` is represented as an empty `Dictionary` (`{}`)
- `Enum::Variant(a, b, c)` is represented as an externally tagged `Dictionary` (`{ "Variant": [a, b, c] }`)

- - -

For ease of use, `ToVariant` and `FromVariant` are also implemented for `Option<T>`, `Result<T, E>` and `Vec<T>`. `Vec<T>` silently drops values for which `from_variant` failed, but the composition with `Option<T>` works nicely, in case it's necessary to know which items failed.

Support for more collections should probably come in the form of iterator support for Godot types, as there is no real way to return or take a `BTreeMap` from Godot, and maps in general can probably have surprising behaviors when their keys are non-primitive (value equality in Rust vs possibly reference equality in other languages).

Fixes #186.

karroffel commented:
This is really cool, I like it! Apart from the code for the generic bounds this is also relatively simple to follow, it's nice :)

toasteater commented:
The code for the generic bounds are initially adapted from `serde_derive` and is still going through changes. I'll refactor it to make it easier to read!

karroffel commented:
Yes I saw the comment that it was taken from serde, so no worries. I think this approach is fine and once this is out of draft phase I would have no problem merging this.

toasteater commented:
Added derive macro for `FromVariant`, and implemented the conversion traits on some common types. I think it's ready now.

Rebased on `master` because of merge conflicts.

toasteater commented:
If we want `Option<T>` to be stricter we can write this instead for `FromVariant`:

```rust
fn from_variant(variant: &Variant) -> Option<Self> {
    if variant.is_nil() {
        Some(None)
    }
    else {
        T::from_variant(variant).map(Some)
    }
}
```

...so it returns `None` for non-`Nil` values that can't be converted to `T`.

karroffel commented on gdnative-core/src/variant.rs:
```diff
@@ -884,4 +929,177 @@ impl FromVariant for Variant {
+
+impl<T> FromVariant for std::marker::PhantomData<T> {
+    fn from_variant(_: &Variant) -> Option<Self> {
+        Some(std::marker::PhantomData)
```
Shouldn't this perform a check and only accept if the `Variant` is null?

toasteater replied:
I think that could work as well.

karroffel commented on gdnative-core/src/variant.rs:
```diff
@@ -884,4 +929,177 @@ impl FromVariant for Variant {
+        let len: usize = arr.len().try_into().ok()?;
+        let mut vec = Vec::with_capacity(len);
+        for idx in 0..len as i32 {
+            if let Some(item) = T::from_variant(arr.get_ref(idx)) {
```
This is a lossy conversion, I don't know what would be the expected behavior honestly.

I would probably expect a heterogeneous array to not be convertible to a homogeneous vector. It would warn you if the caller's code does something weird, it's basically adding run-time checked static typing.

toasteater replied:
You're right about that. It's better to have the user handle values explicitly with `MaybeNot<T>`, if they really want to. The only problem I see is that `godot_wrap_method`'s error message is not very clear about what the expected type is:

```rust
godot_error!("Incorrect argument type {:?} for argument {}",
    _variant.get_type(),
    offset);
```

I'll see if I can improve on that somehow. Maybe `stringify` the argument name and Rust type?

toasteater replied:
Modified the error message to something like: `Incorrect argument type for argument #1 (foo: Vec<Option<Result<i64, u8>>>). Got VariantType::VariantArray (non-primitive types may impose structural checks)`. I think this is enough information.

karroffel commented on gdnative-derive/src/derive_conv_variant.rs:
```diff
@@ -0,0 +1,480 @@
+    let where_clause = &generics.where_clause;
+
+    let result = quote! {
+        #[automatically_derived]
```
I've never seen this and can't find anything in the stdlib docs about this either, mind sharing where this attribute comes from? :smile:

toasteater replied:
I saw it on other derived impls in `cargo-expand` while debugging, and turns out it's a currently undocumented attribute that does a couple of things, and is widely used by derive macros: https://github.com/rust-lang-nursery/reference/issues/532#issuecomment-471086171

I'm not sure if it's really needed, though. Admittedly it's mostly cargo cult here.

karroffel replied:
Hm okay, it doesn't seem like there is much use for it, also Centril expressed concerns about destabilization or clarification of semantics, so I would probably prefer to remove this for now if it doesn't serve much purpose.

toasteater replied:
Removed.
